ibis/backends/conftest.py

```diff
@@ -38,72 +40,111 @@ def data_directory() -> Path:
-    """
-    Convert a backend string to the test configuration class for the backend.
-    """
-    try:
-        backend_package = getattr(ibis, backend_str).__module__
-    except AttributeError:
-        raise ValueError(
-            f'Unknown backend {backend_str}. '
-            f'Known backends: {_get_all_backends()}'
-        )
-
-    conftest = importlib.import_module(f'{backend_package}.tests.conftest')
+    """Convert a backend string to the test class for the backend."""
+    conftest = pytest.importorskip(
+        f"ibis.backends.{backend_str}.tests.conftest"
+    )
     return conftest.TestConf
 
 
-@lru_cache(maxsize=None)
-def _get_backends_to_test(keep=None, discard=None):
-    """
-    Get a list of `TestConf` classes of the backends to test.
-
-    The list of backends can be specified by the user with the
-    `PYTEST_BACKENDS` environment variable.
-
-    - If the variable is undefined or empty, then no backends are returned
-    - Otherwise the variable must contain a space-separated list of backends to
-      test
+def _get_backend_from_parts(parts: tuple[str, ...]) -> str | None:
+    """Return the backend part of a test file's path parts.
 
+    Examples
+    --------
+    >>> _get_backend_from_parts(("/", "ibis", "backends", "sqlite", "tests"))
+    "sqlite"
     """
-    backends_raw = os.environ.get('PYTEST_BACKENDS')
+    try:
+        index = parts.index("backends")
+    except ValueError:
+        return None
+    else:
+        return parts[index + 1]
+
+
+def pytest_ignore_collect(path, config):
+    # get the backend path part
+    #
+    # path is a py.path.local object hence the conversion to Path first
+    backend = _get_backend_from_parts(Path(path).parts)
+    if backend is None or backend not in _get_all_backends():
+        return False
+
+    # we evaluate the marker early so that we don't trigger
+    # an import of conftest files for the backend, which will
+    # import the backend and thus require dependencies that may not
+    # exist
+    #
+    # alternatives include littering library code with pytest.importorskips
+    # and moving all imports close to their use site
+    #
+    # the latter isn't tenable for backends that use multiple dispatch
+    # since the rules are executed at import time
+    mark_expr = config.getoption("-m")
+    # we can't let the empty string pass through, since `'' in s` is `True` for
+    # any `s`; if no marker was passed don't ignore the collection of `path`
+    if not mark_expr:
+        return False
+    expr = _pytest.mark.expression.Expression.compile(mark_expr)
+    # we check the "backend" marker as well since if that's passed
+    # any file matching a backed should be skipped
+    keep = expr.evaluate(lambda s: s in (backend, "backend"))
+    return not keep
+
+
+def pytest_collection_modifyitems(session, config, items):
+    # add the backend marker to any tests are inside "ibis/backends"
+    all_backends = _get_all_backends()
```

Comment:
`_get_all_backends` is ambiguous for me, I'm not sure whether it is a backend name, module or conf class.

Reply:
Yep, changed to `_get_backend_names`.